horreum-backend/src/main/java/io/hyperfoil/tools/horreum/svc/AlertingServiceImpl.java

```diff
@@ -655,33 +674,42 @@ private void runChangeDetection(VariableDAO variable, JsonNode fingerprint, bool
-                          i.testId = (int) tuples[3];
-                          return i;
-                       }).getSingleResult();
-               em.persist(change);
-               Hibernate.initialize(change.dataset.run.id);
-               String testName = TestDAO.<TestDAO>findByIdOptional(variable.testId).map(test -> test.name).orElse("<unknown>");
-               Change.Event event = new Change.Event(ChangeMapper.from(change), testName, DatasetMapper.fromInfo(info), notify);
-               if(mediator.testMode())
-                  Util.registerTxSynchronization(tm, txStatus -> mediator.publishEvent(AsyncEventChannels.CHANGE_NEW, change.dataset.testid, event));
-               mediator.executeBlocking(() -> mediator.newChange(event)) ;
-            });
+            //Only run bulk models on the last datapoint, otherwise run on every datapoint
+            if (model.getType() == ModelType.CONTINOUS || (model.getType() == ModelType.BULK && lastDatapoint)) {
+               try {
+                  model.analyze(dataPoints, detection.config, change -> {
+                     logChangeDetectionMessage(variable.testId, datasetId, PersistentLogDAO.DEBUG,
+                             "Change %s detected using datapoints %s", change, reversedAndLimited(dataPoints));
+                     DatasetDAO.Info info = session
+                             .createNativeQuery("SELECT id, runid as \"runId\", ordinal, testid as \"testId\" FROM dataset WHERE id = ?1", Tuple.class)
+                             .setParameter(1, change.dataset.id)
+                             .setTupleTransformer((tuples, aliases) -> {
+                                DatasetDAO.Info i = new DatasetDAO.Info();
+                                i.id = (int) tuples[0];
+                                i.runId = (int) tuples[1];
+                                i.ordinal = (int) tuples[2];
+                                i.testId = (int) tuples[3];
+                                return i;
+                             }).getSingleResult();
+                     em.persist(change);
+                     Hibernate.initialize(change.dataset.run.id);
+                     String testName = TestDAO.<TestDAO>findByIdOptional(variable.testId).map(test -> test.name).orElse("<unknown>");
+                     Change.Event event = new Change.Event(ChangeMapper.from(change), testName, DatasetMapper.fromInfo(info), notify);
+                     if (mediator.testMode())
+                        Util.registerTxSynchronization(tm, txStatus -> mediator.publishEvent(AsyncEventChannels.CHANGE_NEW, change.dataset.testid, event));
+                     mediator.executeBlocking(() -> mediator.newChange(event));
+                  });
+               } catch (ChangeDetectionException e) {
+                  new ChangeDetectionLogDAO(variable, fingerprint, PersistentLogDAO.ERROR, e.getLocalizedMessage()).persist();
+                  throw new RuntimeException(e);
```

Comment:
I like the idea of logging change detection exceptions in the database, but I think that in this way the _persist_ won't have any effect as we are throwing a `RuntimeException` that IIRC should mark the transaction to be rolled back, thus, ending up in no logs added in the database. am I missing something?